smartlearner/optimizers/optimizer.py

```diff
@@ -12,12 +12,22 @@ def __init__(self, loss):
         self._update_rules = []
         self._param_modifiers = []
 
+        self.graph_updates = OrderedDict()
+        self._directions = None
+
     def append_update_rule(self, update_rule):
         self._update_rules.append(update_rule)
 
     def append_param_modifier(self, param_modifier):
         self._param_modifiers.append(param_modifier)
 
+    @property
+    def directions(self):
+        if self._directions is None:
```

Comment:
I think we have different definitions. For me, graph creation happens when building the cost function to be optimized (`apply` methods of blocks, any call to the `tensor` submodule of *theano*...). Then, the computation graph is compiled to provide outputs and update shared variables.

The final blocks of the graph are built when creating the `Loss` object, which in turn is needed for the creation of an `Optimizer` which is used in the creation of a `Trainer`. So, when tasks are added to the trainer, even though the function isn't compiled, the graph still is complete.

That being said, I agree that dropout should have been applied beforehand. However, nothing prevents someone to make changes to the graph once created. It's actually a pretty big feature of `blocks` : it can add computation in the middle of the graph without having to rebuild the rest of it.

For the problem at hand : caching would work and now that I think of it, since it's not yet compiled, any changes to the graph would be changed in the gradients graph too, so this whole thing isn't really a problem anyway.